### PR TITLE
[WIP] Enable masm compilation under windows

### PR DESCRIPTION
**Pull request by feliwir**
Adresses some issues with the cmake configuration. *.masm files are now compiled with masm as part of the build process. Adresses #474 
TODO:
- [x] Remove include directives from masm assembly files
- [ ] Validate AES-NI support is working

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -289,7 +289,7 @@ if(ENABLE_ASM)
 		endif()
 	elseif(APPLE AND "${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86_64")
 		set(HOST_ASM_MACOSX_X86_64 true)
-	elseif(MSVC AND "${CMAKE_GENERATOR}" MATCHES "Win64" AND FALSE)
+	elseif(MSVC AND "${CMAKE_SYSTEM_PROCESSOR}" MATCHES "AMD64")
```
Comment (feliwir) [inline CMakeLists.txt]: The old fashion of checking for Win64 doesn't work anymore with VS2019

File: crypto/CMakeLists.txt [hunk 2]
```diff
@@ -147,7 +147,7 @@ if(HOST_ASM_MASM_X86_64)
 	add_definitions(-DWHIRLPOOL_ASM)
 	add_definitions(-DOPENSSL_CPUID_OBJ)
 	set(CRYPTO_SRC ${CRYPTO_SRC} ${ASM_X86_64_MASM_SRC})
-	set_property(SOURCE ${ASM_X86_64_MASM_SRC} PROPERTY LANGUAGE C)
+	set_property(SOURCE ${ASM_X86_64_MASM_SRC} PROPERTY LANGUAGE ASM_MASM)
```
Comment (feliwir) [inline crypto/CMakeLists.txt]: Files are getting picked up correctly by VS when setting the language to ASM_NASM

**Comment by busterb** [comment]
To actually build the .asm files under `masm`, we should also pipe them through `cpp` in update.sh as part of the generation process, so there are no `#include` directives in the final product. Let me see if I can whip up a diff.

**Comment by busterb** [comment]
Try adding something like this and run 'update.sh' again:
```
--- a/update.sh
+++ b/update.sh
@@ -167,7 +167,11 @@ gen_asm_stdout() {
        .section .note.GNU-stack,"",%progbits
        #endif
        EOF
-       $MV $3.tmp $3
+       if [ $1 = "masm" ]; then
+               cpp -I./crypto $3.tmp > $3
+       else
+               $MV $3.tmp $3
+       fi
 }
 gen_asm() {
        CC=true perl $asm_src/$2 $1 $3.tmp
@@ -176,7 +180,11 @@ gen_asm() {
        .section .note.GNU-stack,"",%progbits
        #endif
        EOF
-       $MV $3.tmp $3
+       if [ $1 = "masm" ]; then
+               cpp -I./crypto $3.tmp > $3
+       else
+               $MV $3.tmp $3
+       fi
 }
```

That'll preprocess the masm-targeted assembly, though I don't have a machine to test at the moment to verify the output is parsable by it other than validating the `#include` directives are gone and the defines are expanded into constants.

**Comment by feliwir** [comment]
@busterb 
How am i supposed to run that on windows actually?😅 I’ve only used release source trees. When running with mingw i’ve failed.

**Comment by busterb** [comment]
Locally I just run it on Linux, OpenBSD or MacOS and my Windows targets just build from a share. That's the fastest way without having to sort out how to get a posix shell on Windows.

The script also runs fine from cygwin assuming you have installed the prerequisites mentioned in README.md: automake, autoconf, git, libtool, perl. That's what I've used locally in the past, and the environment under which the `dist-win.sh` script was run. I suppose with the above diff you'll also need to at least add the base gcc package to cygwin too to ensure that `cpp` is available.

The `appveyor.yml` file shows a few of the msys64 pacman commands that are used to configure the environment with that tool as well, if that's your preference.

**Comment by busterb** [comment]
Maybe using `WSL` is another choice, though I'm not sure how well file sharing is between it and the base OS until `WSL2` is available.

Anyway, I'm doing a quick test build to get appveyor back on track and to see if the `cpp` update will work on that environment as well. https://github.com/libressl-portable/portable/pull/526

**Comment by busterb** [comment]
Tests in #526 look alright (there was a github timeout, but that's not a big problem). I'll merge your changes in too and we should get some initial results.

**Comment by busterb** [comment]
Nope, needs a little more work (at least when generating on Windows): https://ci.appveyor.com/project/busterb/portable/builds/24616703/job/0pc9rnfdxol9ockc

**Comment by feliwir** [comment]
That’s strange, i’ll check it out when i’m home

**Comment by feliwir** [comment]
So the issue is the OPENSSLDIR variable. All the flags that were set my `add_definition` are passed to MASM aswell, but it seems like the whitespace inside the OPENSSLDIR is causing a failure. E.g. i tried to open 
C:/Program **Files/LibreSSL/etc/ssl**, which is the part after the whitespace. 
Since building always somewhere else is not an option i recommend setting the OPENSSLDIR property on a source level base, instead of defining it globally

**Comment by feliwir** [comment]
Any idea on how to process @busterb ?

**Comment by busterb** [comment]
I just cherry-picked @kinichiro 's fix into #526 . How about we focus on that PR for resolving the remaining masm build issues. Letting it run against travis + appveyor now, and will see how it fares in a bit.

**Comment by busterb** [comment]
Closing this in favor of #526

**Pull request closed without merging**